In [24]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [54]:
# read in the raw data from the github url

df = pd.read_csv("https://raw.githubusercontent.com/suneel0101/lesson-plan/master/crunchbase_monthly_export.csv")

In [55]:
df.head()

permalink                name  \
0   /organization/canal-do-credito    Canal do Credito   
1            /organization/waywire            #waywire   
2  /organization/tv-communications  &TV Communications   
3    /organization/rock-your-paper   'Rock' Your Paper   
4   /organization/in-touch-network   (In)Touch Network   

                       homepage_url  \
0  http://www.canaldocredito.com.br   
1            http://www.waywire.com   
2             http://enjoyandtv.com   
3      http://www.rockyourpaper.org   
4     http://www.InTouchNetwork.com   

                                       category_list          market   \
0               |Credit|Technology|Services|Finance|          Credit    
1         |Entertainment|Politics|Social Media|News|   Entertainment    
2                                            |Games|           Games    
3                             |Publishing|Education|       Education    
4  |Electronics|Guides|Coffee|Restaurants|Music|i...            Apps    

   funding_total_usd      status country_code state_code          region  \
0            750,000         NaN          BRA        NaN  Rio de Janeiro   
1          1,750,000    acquired          USA         NY   New York City   
2          4,000,000   operating          USA         CA     Los Angeles   
3             40,000   operating          EST        NaN         Tallinn   
4          1,500,000   operating          GBR        NaN          London   

             city  funding_rounds founded_at founded_month founded_quarter  \
0  Belo Horizonte               1        NaN           NaN             NaN   
1        New York               1     6/1/12       2012-06         2012-Q2   
2     Los Angeles               2        NaN           NaN             NaN   
3         Tallinn               1   10/26/12       2012-10         2012-Q4   
4          London               1     4/1/11       2011-04         2011-Q2   

   founded_year first_funding_at last_funding_at  Unnamed: 18  
0           NaN           1/1/10          1/1/10          NaN  
1        2012.0          6/30/12         6/30/12          NaN  
2           NaN           6/4/10         9/23/10          NaN  
3        2012.0           8/9/12          8/9/12          NaN  
4        2011.0           4/1/11          4/1/11          NaN

#### 1. Clean the data. Use .apply and lambda functions where you can

In [57]:
# rename the columns to remove spaces and extra characters
df.columns = ['permalink','name','homepage_url', 'category_list', 'market', 'funding_total_usd', 'status', 'country_code', 'state_code', 'region', 'city', 'funding_rounds', 'founded_at', 'founded_month','founded_quarter','founded_year','first_funding_at', 'last_funding_at','undefined']
df.head()

permalink                name  \
0   /organization/canal-do-credito    Canal do Credito   
1            /organization/waywire            #waywire   
2  /organization/tv-communications  &TV Communications   
3    /organization/rock-your-paper   'Rock' Your Paper   
4   /organization/in-touch-network   (In)Touch Network   

                       homepage_url  \
0  http://www.canaldocredito.com.br   
1            http://www.waywire.com   
2             http://enjoyandtv.com   
3      http://www.rockyourpaper.org   
4     http://www.InTouchNetwork.com   

                                       category_list           market  \
0               |Credit|Technology|Services|Finance|          Credit    
1         |Entertainment|Politics|Social Media|News|   Entertainment    
2                                            |Games|           Games    
3                             |Publishing|Education|       Education    
4  |Electronics|Guides|Coffee|Restaurants|Music|i...            Apps    

  funding_total_usd     status country_code state_code          region  \
0          750,000         NaN          BRA        NaN  Rio de Janeiro   
1        1,750,000    acquired          USA         NY   New York City   
2        4,000,000   operating          USA         CA     Los Angeles   
3           40,000   operating          EST        NaN         Tallinn   
4        1,500,000   operating          GBR        NaN          London   

             city  funding_rounds founded_at founded_month founded_quarter  \
0  Belo Horizonte               1        NaN           NaN             NaN   
1        New York               1     6/1/12       2012-06         2012-Q2   
2     Los Angeles               2        NaN           NaN             NaN   
3         Tallinn               1   10/26/12       2012-10         2012-Q4   
4          London               1     4/1/11       2011-04         2011-Q2   

   founded_year first_funding_at last_funding_at  undefined  
0           NaN           1/1/10          1/1/10        NaN  
1        2012.0          6/30/12         6/30/12        NaN  
2           NaN           6/4/10         9/23/10        NaN  
3        2012.0           8/9/12          8/9/12        NaN  
4        2011.0           4/1/11          4/1/11        NaN

In [16]:
df.describe()

funding_rounds  founded_year  undefined
count    44950.000000  35226.000000        0.0
mean         1.681135   2007.083461        NaN
std          1.269198      7.522161        NaN
min          1.000000   1900.000000        NaN
25%          1.000000           NaN        NaN
50%          1.000000           NaN        NaN
75%          2.000000           NaN        NaN
max         16.000000   2014.000000        NaN

In [52]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44950 entries, 0 to 44949
Data columns (total 19 columns):
permalink            44950 non-null object
name                 44949 non-null object
homepage_url         41914 non-null object
category_list        40997 non-null object
market               40997 non-null object
funding_total_usd    0 non-null float64
status               42208 non-null object
country_code         40805 non-null object
state_code           27773 non-null object
region               39615 non-null object
city                 40109 non-null object
funding_rounds       44950 non-null int64
founded_at           35292 non-null object
founded_month        35226 non-null object
founded_quarter      35226 non-null object
founded_year         35226 non-null float64
first_funding_at     44950 non-null object
last_funding_at      44950 non-null object
undefined            0 non-null float64
dtypes: float64(3), int64(1), object(15)
memory usage: 6.5+ MB


In [21]:
df.get_dtype_counts()

float64     2
int64       1
object     16
dtype: int64

In [33]:
df.columns

Index([u'permalink', u'name', u'homepage_url', u'category_list', u'market',
       u'funding_total_usd', u'status', u'country_code', u'state_code',
       u'region', u'city', u'funding_rounds', u'founded_at', u'founded_month',
       u'founded_quarter', u'founded_year', u'first_funding_at',
       u'last_funding_at', u'undefined'],
      dtype='object')

In [79]:
# convert the funding, year founded, and funding_rounds columns to integers
#df['funding_total_usd'] = df.funding_total_usd.str.replace(',','')
#df['funding_total_usd'] = df['funding_total_usd'].astype('int')
#df['funding_total_usd'] = df.apply(lambda x: int(x))
#df.funding_total_usd = df.funding_total_usd.map(int)
#df['funding_total_usd'] = df['funding_total_usd'].apply(lambda x: int(x) if str.isdigit(x) else np.nan)
df['funding_total_usd'] = df.funding_total_usd.str.replace('-',np.nan)

#print (df[ pd.to_numeric(df['funding_total_usd'], errors='coerce').isnull()])


In [78]:
df.head(400)

permalink                     name  \
0           /organization/canal-do-credito         Canal do Credito   
1                    /organization/waywire                 #waywire   
2          /organization/tv-communications       &TV Communications   
3            /organization/rock-your-paper        'Rock' Your Paper   
4           /organization/in-touch-network        (In)Touch Network   
5                    /organization/n-plusn                   #NAME?   
6               /organization/club-domains            .Club Domains   
7               /organization/fox-networks            .Fox Networks   
8                    /organization/0-6-com                  0-6.com   
9           /organization/004-technologies         004 Technologies   
10                    /organization/0xdata                   0xdata   
11               /organization/1-2-3-listo              1,2,3 Listo   
12                   /organization/1-4-all                  1-4 All   
13             /organization/1-800-dentist            1-800-DENTIST   
14             /organization/1-800-doctors            1-800-DOCTORS   
15          /organization/1-618-technology         1.618 Technology   
16               /organization/10-20-media              10-20 Media   
17                /organization/1000-corks               1000 Corks   
18              /organization/1000-markets             1000 Markets   
19         /organization/1000jobboersen-de        1000jobboersen.de   
20              /organization/1000memories             1000memories   
21           /organization/1000museums-com          1000museums.com   
22                /organization/1001-menus               1001 Menus   
23                   /organization/1006-tv                  1006.tv   
24                  /organization/100du-tv                 100du.tv   
25                  /organization/100e-com                 100e.com   
26                   /organization/100plus                  100Plus   
27                  /organization/1010data                 1010data   
28              /organization/10bestthings             10BestThings   
29                     /organization/10sec                    10sec   
..                                     ...                      ...   
370                   /organization/8hands                   8hands   
371  /organization/8minutenergy-renewables  8minutenergy Renewables   
372                /organization/8th-story                8th Story   
373              /organization/alvenda-inc                8thBridge   
374                  /organization/8tracks            8tracks Radio   
375                    /organization/8trip                    8Trip   
376                /organization/8villages                8villages   
377                  /organization/8x8-inc                  8x8 Inc   
378              /organization/908-devices              908 Devices   
379       /organization/90sec-technologies       90sec Technologies   
380        /organization/91-boyuan-wireles        91 Boyuan Wireles   
381                  /organization/91-golf                  91 Golf   
382              /organization/91-wireless              91 Wireless   
383                 /organization/911-pets                 911 Pets   
384           /organization/9158-julur-com           9158 Julur.com   
385             /organization/91datong-com             91datong.com   
386                /organization/91jinrong                91JinRong   
387            /organization/99-fahrenheit            99 Fahrenheit   
388                    /organization/99-co                    99.co   
389                   /organization/99bill                   99Bill   
390         /organization/99degrees-custom         99degrees Custom   
391                /organization/99designs                99designs   
392                /organization/99dresses                99dresses   
393                 /organization/99inn-cc                 99inn.cc   
394               /organization/99presents               

In [ ]:
# convert category_list so that each cell is a list of categories

In [ ]:
# use value_counts on market, status, and country_code to check for any repeats or misnamed entries

In [ ]:
# BONUS: convert the date columns to appropriate datetime objects

#### 2. Using isnull(), take a subset of the dataframe that has no founded at date

#### 3. Create a new column, founded_year_x, that's a copy of founded_year. Then, replace all the missing values with either the mean , mode, or median of that column. 
Add a note explaining which statistic you picked and why:

#### 4. Take a subset of US-based companies using boolean selection

#### 5. Create a pivot table showing total, mean, and median funding amount by year

#### 6. Create a pivot table showing total, mean and median funding amount by market

#### 7. CHALLENGE: Using matplotlib, create a scatter plot showing funding amount on the y-axis, and first_funding_date on the x-axis
(first_funding_date will need to be a timestamp)

#### 8. Here's the fun part. Create a hypothesis based on your data that you can reasonably test. 
For example: "Credit companies based in the US that only get 1 round of funding tend to receive more funding than companies based outside the US"



When you have your hypothesis, write it below, and Slack it to the instructors so we can check in. 

#### 9. Test your hypothesis. 

You'll need to evaluate your hypothesis using a t-test (see Monday's lecture 1.3), so make sure you have enough data compare samples. If your original hypothesis is too narrow, you may need to broaden the scope.

#### 10. CHALLENGE
- Plot the distribution of both samples
- Explain your results in one short sentence. 
- Explain your result in a slighly longer technical paragraph, touching on the statistics you presented. 
- Subset your findings: if you looked at a certain market in a certain geography, break it down by another variable (e.g number of funding rounds) and compare those samples. 